db/db.go

```diff
@@ -76,20 +78,70 @@ func Init(config config.DBConfig, debug bool) *Database {
 	if debug {
 		db.Config.Logger.LogMode(logger.Info)
 	}
-	return &Database{db}
+
+	d := &Database{db}
+	if err = d.MigrateLineage(); err != nil {
+		log.Fatalf("Lineage migration failed: %v\n", err)
+	}
+
+	return d
+}
+
+// MigrateLineage is a migration function to update db and its data to the
+// new lineage db scheme. It will update State table data, delete "lineage" column
+// and add corresponding Lineage entries
+func (db *Database) MigrateLineage() error {
+	if db.Migrator().HasColumn(&types.State{}, "lineage") {
+		states := db.ListStates()
+		for _, stPath := range states {
+			// Recover State from db for update
+			var st types.State
+			res := db.First(&st, types.State{Path: stPath})
+			if errors.Is(res.Error, gorm.ErrRecordNotFound) {
+				return fmt.Errorf("State not found in db")
+			}
+
+			if err := db.UpdateState(st); err != nil {
+				return fmt.Errorf("Failed to update %s state during lineage migration: %v", stPath, err)
+			}
+		}
+
+		// Custom migration rules
+		if err := db.Migrator().DropColumn(&types.State{}, "lineage"); err != nil {
+			return fmt.Errorf("Failed to drop lineage column during migration: %v", err)
+		}
+	}
+
+	return nil
 }
 
 type attributeValues map[string]interface{}
 
 func (db *Database) stateS3toDB(sf *statefile.File, path string, versionID string) (st types.State) {
 	var version types.Version
 	db.First(&version, types.Version{VersionID: versionID})
-	st = types.State{
-		Path:      path,
-		Version:   version,
-		TFVersion: sf.TerraformVersion.String(),
-		Serial:    int64(sf.Serial),
-		Lineage:   sf.Lineage,
+
+	// Check if the associated lineage is already present in lineages table
+	// If so, it recovers its ID otherwise it inserts it at the same time as the state
+	var lineage types.Lineage
+	if errors.Is(db.First(&lineage, types.Lineage{Value: sf.Lineage}).Error, gorm.ErrRecordNotFound) {
```

Comment:
Maybe it's just me, but I find this line a bit loaded. Can we split into:

```suggestion
      err := db.First(&lineage, types.Lineage{Value: sf.Lineage}).Error
	if errors.Is(err, gorm.ErrRecordNotFound) {
```

Reply:
Arguably, this should never happen this the migration is done when Terraboard is launched. Or am I missing something?

Reply:
> Arguably, this should never happen this the migration is done when Terraboard is launched. Or am I missing something?

Precisely, this condition was introduced by the arrival of the new lineage table. Indeed, when we insert a state there are two possible scenarios:
- The associated lineage was already inserted in the lineages table previously by another state. So, we just recover the lineage id and set the foreign key directly.
- The lineage isn't in the lineages table, so we set the lineage structure in the state one so that gorm takes care of inserting it at the same time as the state